x-pack/filebeat/input/httpjson/request.go

```diff
@@ -163,21 +181,106 @@ func newRequester(
+	req, err := rf.newHTTPRequest(stdCtx, trCtx)
 	if err != nil {
-		return fmt.Errorf("failed to create http request: %w", err)
+		return nil, fmt.Errorf("failed to create http request: %w", err)
 	}
-
 	httpResp, err := r.client.do(stdCtx, trCtx, req)
 	if err != nil {
-		return fmt.Errorf("failed to execute http client.Do: %w", err)
+		return nil, fmt.Errorf("failed to execute http client.Do: %w", err)
+	}
+	return httpResp, nil
+}
+
+// generateNewUrl returns new url value using replacement from oldUrl with str ids
+func generateNewUrl(replacement string, oldUrl string, str []byte) (url.URL, error) {
+	reg, err := regexp.Compile(replacement)
+	if err != nil {
+		return url.URL{}, fmt.Errorf("failed to create regex on provided value: %w", err)
+	}
+	newUrl, err := url.Parse(string(reg.ReplaceAll([]byte(oldUrl), str)))
+	if err != nil {
+		return url.URL{}, fmt.Errorf("failed to replace value in url: %w", err)
+	}
+	return *newUrl, nil
+
+}
+
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
+	var (
+		httpResp       *http.Response
+		ArrayhttpResp  []*http.Response
+		finalHttpResps []*http.Response
+		st             []string
+		err            error
+		b              []byte
+		url            string
+	)
+	// perform each requestFactory
+	for i, rf := range r.requestFactory {
+		// iterate over collected ids from last response
+		if i != 0 {
+			url = rf.url.String()
+			// perform request over collected ids
+			for _, s := range st {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, url, []byte(s))
+				if err != nil {
+					return fmt.Errorf("failed to generate new URL: %w", err)
+				}
+
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute http client.Do: %w", err)
+				}
+				// store data according to response type
+				if i+1 == len(r.requestFactory) && len(st) != 0 {
+					finalHttpResps = append(finalHttpResps, httpResp)
+				} else {
+					ArrayhttpResp = append(ArrayhttpResp, httpResp)
+				}
+			}
+		} else {
+			// perform and store regular call responses
+			httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+			if err != nil {
+				return fmt.Errorf("failed to execute http client.Do: %w", err)
+			}
+			if len(r.requestFactory) <= 1 {
+				finalHttpResps = append(finalHttpResps, httpResp)
+			} else {
+				ArrayhttpResp = append(ArrayhttpResp, httpResp)
+			}
+		}
+
+		// parse and store ids from collected responses
+		if i+1 < len(r.requestFactory) {
+			st = nil
+			// collect ids from all responses
+			for _, resp := range ArrayhttpResp {
+				if resp.Body != nil {
+					b, err = ioutil.ReadAll(resp.Body)
+					if err != nil {
+						return err
+					}
+				}
+				// Restore the io.ReadCloser to its original state
+				resp.Body = ioutil.NopCloser(bytes.NewBuffer(b))
```

Comment:
I admin I got lost here, why is necessary to reset it? Is there an option to simplify this logic?

Reply:
Why I did it is because I don't want to store buffered value in the same state.